doc/api/vm.md

```diff
@@ -459,47 +457,56 @@ const bar = new vm.SourceTextModule(`
 //
 // "Link" the imported dependencies of this Module to it.
 //
-// The provided linking callback (the "linker") accepts two arguments: the
-// parent module (`bar` in this case) and the string that is the specifier of
-// the imported module. The callback is expected to return a Module that
-// corresponds to the provided specifier, with certain requirements documented
-// in `module.link()`.
-//
-// If linking has not started for the returned Module, the same linker
-// callback will be called on the returned Module.
+// Obtain the requested dependencies of a SourceTextModule by
+// `sourceTextModule.moduleRequests` and resolve them.
 //
 // Even top-level Modules without dependencies must be explicitly linked. The
-// callback provided would never be called, however.
-//
-// The link() method returns a Promise that will be resolved when all the
-// Promises returned by the linker resolve.
+// array passed to `sourceTextModule.linkRequestedModules(modules)` can be
+// empty, however.
 //
-// Note: This is a contrived example in that the linker function creates a new
+// Note: This is a contrived example in that the linker creates a new
 // "foo" module every time it is called. In a full-fledged module system, a
 // cache would probably be used to avoid duplicated modules.
 
-async function linker(specifier, referencingModule) {
-  if (specifier === 'foo') {
-    return new vm.SourceTextModule(`
-      // The "secret" variable refers to the global variable we added to
-      // "contextifiedObject" when creating the context.
-      export default secret;
-    `, { context: referencingModule.context });
+const moduleMap = new Map([
+  ['root', rootModule],
+]);
 
-    // Using `contextifiedObject` instead of `referencingModule.context`
-    // here would work as well.
-  }
-  throw new Error(`Unable to resolve dependency: ${specifier}`);
+function linker(module) {
+  const requestedModules = module.moduleRequests.map((request) => {
+    // In a full-fledged module system, the linker would resolve the
+    // module with the module cache key `[specifier, attributes]`.
+    // In this example, we just use the specifier as the key.
+    const specifier = request.specifier;
+
+    let requestedModule = moduleMap.get(specifier);
+    if (requestedModule === undefined) {
+      requestedModule = new vm.SourceTextModule(`
+        // The "secret" variable refers to the global variable we added to
+        // "contextifiedObject" when creating the context.
+        export default secret;
+      `, { context: referencingModule.context });
+      moduleMap.set(specifier, linkedModule);
+      // Resolve the dependencies of the new module as well.
+      linker(requestedModule);
+    }
+
+    return requestedModule;
+  });
+
+  module.linkRequestedModules(requestedModules);
```

Comment:
This is a great design! But I would like to suggest two simplifications:

1. Call it `module.linkRequests` instead of `module.linkRequestedModules`
2. Support an object sugar form like `module.linkRequests({ 'foo': moduleMap.get('foo') })`. While this form does not support import attributes, it handles the ergonomic common usage.

In future we likely want to support partial linking as well with host fallbacks for non-explicit linkage. And while both the array form and the object form in (2) can support partial linkage by providing `undefined`, it is still more ergonomic as well in the case of (2).